data_analysis/correction_framework.py

```diff
@@ -0,0 +1,290 @@
+                    continue
+                self.min_records.append(m_rec)
+                trans_str = n_file.readline()
+
+        with open(morning_file) as m_file:
+            trans_str = m_file.readline()
+            while len(trans_str) != 0:
+                data = trans_str.split(";")
+                data[0] = data[0].replace("T", " ")
+                data[1] = data[1].replace("T", " ")
+                m_rec = MinuteRecord(
+                    Time(data[0]),
+                    Time(data[1]),
+                    float(data[2]),
+                    float(data[3]),
+                    float(data[4]),
+                    data[5],
+                )
+                if m_rec.utc_time > self.sunrise:
+                    break
+                self.min_records.append(m_rec)
+                trans_str = m_file.readline()
+
+        self.valid_for_use = True
+        self.exclusion_reason = "None"
+
+    # grade quality of as based on various night conditions
+    def grade_data(self):
+        # taking the average night to be ten hours (600 minutes), if more than
+        # three hours (180 minutes) are missing, invalidate the data based on
+        # insufficient records
+        if len(self.min_records) < 420:
+            self.valid_for_use = False
+            self.exclusion_reason = "Insufficient Records for Night"
+
+        # future TODO: invalidate based on weather conditions
+        # potential future TODO: grade data on a scale (good, poor, unusable,
+        #  etc) instead of just valid or invalid
+
+    # given a function on a float, updates all temperature, frequency, or
+    # voltage values
+    def apply_correction(self, field_to_update: str, funct):
+        if field_to_update == "TEMP":
+            for rec in self.min_records:
+                rec.temperature = funct(rec.temperature)
+        if field_to_update == "FREQ":
+            for rec in self.min_records:
+                rec.frequency = funct(rec.frequency)
+        if field_to_update == "VOLT":
+            for rec in self.min_records:
+                rec.voltage = funct(rec.voltage)
+        else:
+            print("error: field not recognized/field should not be updated")
+
+    # export night record to a new file
+    def rec_export(self, filename: str):
+        with open(filename, "x") as new_file:
+            if not self.valid_for_use:
+                new_file.write("INVALID\n" + self.exclusion_reason)
+            else:
+                vals = [
+                    self.night_date.strftime(PROJECT_DATE_FORMAT),
+                    self.morning_date.strftime(PROJECT_DATE_FORMAT),
+                    self.sunrise.strftime(PROJECT_DATE_FORMAT),
+                    self.sunset.strftime(PROJECT_DATE_FORMAT),
+                    self.astronomical_twilight.strftime(PROJECT_DATE_FORMAT),
+                    str(self.moon_illumination),
+                ]
+                new_file.write("\n".join(vals) + "\n")
```

Comment:
If you are writing a line with a `\n` character at the end, you could also do this:
```python
print("\n".join(vals), file=file_output)
```